hydra-node/src/Hydra/PersistentQueue.hs

```diff
@@ -0,0 +1,148 @@
+  pure PersistentQueue{queue, nextIx, directory = path}
+ where
+  loadExisting queue = do
+    paths <- liftIO $ listDirectory path
+    case sort $ mapMaybe readMaybe paths of
+      [] -> pure 0
+      idxs -> do
+        forM_ idxs $ \(idx :: Natural) -> do
+          bs <- readFileLBS (path </> show idx)
+          case eitherDecode' bs of
+            Left err ->
+              fail $ "Failed to decode item: " <> show err
+            Right item ->
+              atomically $ writeTBQueue queue (idx, item)
+        pure $ List.last idxs
+
+-- | Write a value to the queue, blocking if the queue is full.
+writeDurablePersistentQueue :: (ToCBOR a, MonadSTM m, MonadIO m) => PersistentQueue m a -> a -> m ()
+writeDurablePersistentQueue PersistentQueue{queue, nextIx, directory} item = do
+  next <- atomically $ do
+    next <- readTVar nextIx
+    modifyTVar' nextIx (+ 1)
+    pure next
+  writeBinaryFileDurable (directory </> show next) $ serialize' item
+  atomically $ writeTBQueue queue (next, item)
+
+-- | Write a value to the queue, blocking if the queue is full.
+writePersistentQueue :: (ToCBOR a, MonadSTM m, MonadIO m) => PersistentQueue m a -> a -> m ()
+writePersistentQueue PersistentQueue{queue, nextIx, directory} item = do
+  next <- atomically $ do
+    next <- readTVar nextIx
+    modifyTVar' nextIx (+ 1)
+    pure next
+
+  liftIO $ createDirectoryIfMissing True directory
+  writeFileBS (directory </> show next) $ serialize' item
+  atomically $ writeTBQueue queue (next, item)
+
+-- | Write a value to the queue, blocking if the queue is full.
+writePersistentQueueJson :: (ToJSON a, MonadSTM m, MonadIO m) => PersistentQueue m a -> a -> m ()
+writePersistentQueueJson PersistentQueue{queue, nextIx, directory} item = do
+  next <- atomically $ do
+    next <- readTVar nextIx
+    modifyTVar' nextIx (+ 1)
+    pure next
+
+  liftIO $ createDirectoryIfMissing True directory
+  writeFileBS (directory </> show next) $
+    toStrict (encode item)
+  atomically $ writeTBQueue queue (next, item)
+
+-- | Get the next value from the queue without removing it, blocking if the
+-- queue is empty.
+peekPersistentQueue :: MonadSTM m => PersistentQueue m a -> m (Natural, a)
+peekPersistentQueue PersistentQueue{queue} = do
+  atomically (peekTBQueue queue)
+
+-- | Remove an element from the queue if it matches the given item. Use
+-- 'peekPersistentQueue' to wait for next items before popping it.
+popPersistentQueue :: (MonadSTM m, MonadIO m, Eq a) => PersistentQueue m a -> a -> m ()
+popPersistentQueue PersistentQueue{queue, directory} item = do
+  popped <- atomically $ do
+    (ix, next) <- peekTBQueue queue
+    if next == item
+      then readTBQueue queue $> Just ix
+      else pure Nothing
+  case popped of
+    Nothing -> pure ()
+    Just index -> do
+      let path = directory </> show index
+      fileExists <- liftIO $ doesPathExist path
+      when fileExists $
+        liftIO $
+          removeFile path
```

Comment:
Never ask for permission, but for forgiveness .. also when doing file IO.

While this might work, I'd suggest to just remove the file and ignore not exists errors. This is generally good practice because if it does not exist in one moment, it might exist in the next.. or the other way around. It should not be the case here, but just wanted to recommend that practice.

However, if the file does not exist here, something is off .. maybe even exceptionally off and we should fail here?

Reply:
The `NodeSpec` tests were wired incorrectly (temp dir was removed before returning tests results). Now all works without checking for file existence.